MonoGame.Framework/Media/VideoPlayer.WMS.cs

```diff
@@ -59,102 +87,132 @@ private void PlatformInitialize()
+            WaitForInternalStateChange(InternalState.Paused);
         }
 
         private void PlatformPlay()
         {
-            // Cleanup the last song first.
+            System.Diagnostics.Debug.WriteLine("PlatformPlay");
+            // Cleanup the last video first.
             if (State != MediaState.Stopped)
             {
-                _session.Stop();
-                _session.ClearTopologies();
-                _session.Close();
-                if (_volumeController != null)
-                {
-                    _volumeController.Dispose();
-                    _volumeController = null;
-                }
-                _clock.Dispose();
+                PlatformStop();
             }
 
-            //create the callback if it hasn't been created yet
+            CreateTexture();
+
+            // Create the callback if it hasn't been created yet
             if (_callback == null)
             {
                 _callback = new Callback(this);
                 _session.BeginGetEvent(_callback, null);
             }
 
-            // Set the new song.
+            // Set the new video.
+            _internalState = InternalState.WaitingForSessionStart;
             _session.SetTopology(SessionSetTopologyFlags.Immediate, _currentVideo.Topology);
 
-            // Get the clock.
-            _clock = _session.Clock.QueryInterface<PresentationClock>();
-
-            // Start playing.
-            var varStart = new Variant();
-            _session.Start(null, varStart);
+            WaitForInternalStateChange(InternalState.Playing);
         }
 
         private void PlatformResume()
         {
+            _internalState = InternalState.WaitingForSessionStart;
             _session.Start(null, null);
+            WaitForInternalStateChange(InternalState.Playing);
         }
 
         private void PlatformStop()
         {
-            _session.ClearTopologies();
-            _session.Stop();
-            _session.Close();
-            if (_volumeController != null)
+            System.Diagnostics.Debug.WriteLine("PlatformStop");
+            if (State == MediaState.Playing)
             {
-                _volumeController.Dispose();
-                _volumeController = null;
+                _internalState = InternalState.WaitingForSessionStop;
+                _session.Stop();
+                WaitForInternalStateChange(InternalState.Stopped);
             }
-            _clock.Dispose();
-            _clock = null;
+            else
+            {
+                _internalState = InternalState.Stopped;
+            }
+            System.Diagnostics.Debug.WriteLine("PlatformStopped");
+        }
+
+        bool WaitForInternalStateChange(InternalState expectedState, int milliseconds = defaultTimeoutMs)
+        {
+            var timer = System.Diagnostics.Stopwatch.StartNew();
+            while (_internalState != expectedState)
+            {
+#if WINRT
+                Task.Delay(1).Wait();
```

Comment:
`Task.Delay` actually generates garbage. An alternative which works just as well and doesn't generate garbage is `Monitor.Wait`. PR #4207 has an example of this (specifically commit 0a18376).